admin/permissions.go

```diff
@@ -308,3 +313,51 @@ func UnionProjectRoles(a *adminv1.ProjectPermissions, b *database.ProjectRole) *
 		ManageBookmarks:            a.ManageBookmarks || b.ManageBookmarks,
 	}
 }
+
+// UnionUserProjectRoles merges a user's project role's permissions into the given permissions object.
+func UnionUserProjectRoles(a *adminv1.ProjectPermissions, b *database.UserProjectRole) *adminv1.ProjectPermissions {
+	var mergedResources []*adminv1.ResourceName
+	seen := make(map[string]struct{})
+	for _, res := range a.Resources {
+		key := res.Type + "|" + res.Name
+		seen[key] = struct{}{}
+		mergedResources = append(mergedResources, res)
+	}
+	for _, res := range b.Resources {
+		key := res.Type + "|" + res.Name
+		if _, ok := seen[key]; !ok {
+			seen[key] = struct{}{}
+			mergedResources = append(mergedResources, &adminv1.ResourceName{
+				Type: res.Type,
+				Name: res.Name,
+			})
+		}
+	}
+
+	return &adminv1.ProjectPermissions{
+		Admin:                      a.Admin || b.Admin,
+		ReadProject:                a.ReadProject || b.ReadProject,
+		ManageProject:              a.ManageProject || b.ManageProject,
+		ReadProd:                   a.ReadProd || b.ReadProd,
+		ReadProdStatus:             a.ReadProdStatus || b.ReadProdStatus,
+		ManageProd:                 a.ManageProd || b.ManageProd,
+		ReadDev:                    a.ReadDev || b.ReadDev,
+		ReadDevStatus:              a.ReadDevStatus || b.ReadDevStatus,
+		ManageDev:                  a.ManageDev || b.ManageDev,
+		ReadProvisionerResources:   a.ReadProvisionerResources || b.ReadProvisionerResources,
+		ManageProvisionerResources: a.ManageProvisionerResources || b.ManageProvisionerResources,
+		ReadProjectMembers:         a.ReadProjectMembers || b.ReadProjectMembers,
+		ManageProjectMembers:       a.ManageProjectMembers || b.ManageProjectMembers,
+		ManageProjectAdmins:        a.ManageProjectAdmins || b.ManageProjectAdmins,
+		CreateMagicAuthTokens:      a.CreateMagicAuthTokens || b.CreateMagicAuthTokens,
+		ManageMagicAuthTokens:      a.ManageMagicAuthTokens || b.ManageMagicAuthTokens,
+		CreateReports:              a.CreateReports || b.CreateReports,
+		ManageReports:              a.ManageReports || b.ManageReports,
+		CreateAlerts:               a.CreateAlerts || b.CreateAlerts,
+		ManageAlerts:               a.ManageAlerts || b.ManageAlerts,
+		CreateBookmarks:            a.CreateBookmarks || b.CreateBookmarks,
+		ManageBookmarks:            a.ManageBookmarks || b.ManageBookmarks,
+		FullyResourceRestricted:    a.FullyResourceRestricted && len(b.Resources) > 0,
```

Comment:
I feel a little uneasy by the importance attached to whether the list is empty or not. 

I also think we may need to support having user members who don't have access to any resources. (I.e. not automatically remove the member just because their access was removed.)

Not sure, but what do you think of this idea:
- You only have access to the resources in `b.Resources`
- We add a resource type to represent "all access", e.g. `ResourceName{Kind: "*"}` or something like that. (This can be added to all existing users with a migration)

That would remove the need to handle empty lists or pass a separate field/permission around